Add scalar systems to linearFV version of SIMPLE

GiudGiud opened the pull request:
- [x] kernels
- [x] solve object attributes
- [x] solve the systems
- [x] tests
- [x] documentation

GiudGiud committed:
Add scalar advection kernels
refs #29165

GiudGiud committed:
Add test for 2 scalar advection
refs #29165
Add docs

GiudGiud committed:
Move solve attributes for passive scalar to base SIMPLE class
refs #29165

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2567354/), step Docs: sync website on a75c326 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29156/site)

This comment will be updated on new commits.

GiudGiud committed:
Add support for passive scalar solves in SIMPLE
refs #29165

GiudGiud committed:
Add a base class for scalar transport WCNSFV physics
refs #29175

GiudGiud committed:
Add a physics class for the linear FV discretization of the scalar advection equation
refs #29175

GiudGiud committed:
Add test for scalar advection physics
add doco
refs #29175

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2572029/), step Clang format on 6ab0036 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format efd9ea4e203a5628c080244d6e57becc3ef322b2`

GiudGiud commented:
Should be ready for review. I'm happy to remove the 2 commits adding a Physics if this is distracting us from the critical path of adding these linear capabilities

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2572173/), step Generate coverage on 30262af wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>efd9ea</th>
      <th colspan=4>#29156 30262a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.14%</td>
      <td>85.14%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>107573</td>
      <td>107574</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18776</td>
      <td>18775</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29156/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29156/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>efd9ea</th>
      <th colspan=4>#29156 30262a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.82%</td>
      <td>84.72%</td>
      <td>-0.10%</td>
      <td>82.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17329</td>
      <td>17484</td>
      <td>+155</td>
      <td>272</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3102</td>
      <td>3153</td>
      <td>+51</td>
      <td>57</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29156/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29156/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29156/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 82.67% is less than the suggested 90.0%



This comment will be updated on new commits.

grmnptr commented on modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:
```diff
@@ -0,0 +1,23 @@
+\int\limits_{V_C} \nabla \cdot \left(\rho\vec{u} C_i \right) dV \approx \sum\limits_f (\rho \vec{u}\cdot \vec{n})_{RC} C_if |S_f| \,
+\end{equation}
+
+where $C_if$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
```
```suggestion
where $C_{if}$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
```

grmnptr commented on modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:
```diff
@@ -0,0 +1,23 @@
+
+where $C_if$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
+
+The face mass flux $(\rho \vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```
```suggestion
The face mass flux $(\vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```

grmnptr commented on modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:
```diff
@@ -0,0 +1,23 @@
+# LinearFVScalarAdvection
```
The scalar advection schemes don't use the density. We will need some modifications in the RC object to be able to obtain the face velocities (volumetric flow rates).

GiudGiud replied:
very true. this should be fixed now. That was a big oversight on my part, what was I thinking

grmnptr commented on modules/navier_stokes/doc/content/source/physics/WCNSLinearFVScalarTransportPhysics.md:
```diff
@@ -0,0 +1,39 @@
+
+- [LinearFVScalarAdvection.md] for the scalar advection term
+- [LinearFVDiffusion.md] for the scalar diffusion term
+- [LinearFVSource.md] for the reaction terms
```
```suggestion
- [LinearFVReaction.md] for the reaction terms
```

grmnptr commented on modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h:
```diff
@@ -0,0 +1,55 @@
+#include "LinearFVAdvectionDiffusionBC.h"
+
+/**
+ * An advection kernel that implements the advection term for the enthalpy in the
+ * energy equation.
```
```suggestion
 * An advection kernel that implements the advection term for the passive scalar transport equation.
```

grmnptr commented on modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h:
```diff
@@ -0,0 +1,55 @@
+
+protected:
+  /// The Rhie-Chow user object that provides us with the face velocity
+  const RhieChowMassFlux & _mass_flux_provider;
```
Maybe we can just divide the mass flux by the face rho. Either in the RC object or in the kernel.

grmnptr commented on modules/navier_stokes/include/physics/WCNSLinearFVScalarTransportPhysics.h:
```diff
@@ -0,0 +1,45 @@
+
+/**
+ * Creates all the objects needed to solve the Navier Stokes scalar transport equations
+ * using the nonlinear finite volume weakly-compressible discretization (WCNSFV)
```
```suggestion
 * using the linear finite volume weakly-compressible discretization (WCNSFV)
```

grmnptr commented on modules/navier_stokes/include/physics/WCNSLinearFVScalarTransportPhysics.h:
```diff
@@ -0,0 +1,45 @@
+  WCNSLinearFVScalarTransportPhysics(const InputParameters & parameters);
+
+private:
+  // TODO Rename to linear or solver?
```
Solver sounds good!

grmnptr commented on modules/navier_stokes/src/executioners/SIMPLESolve.C:
```diff
@@ -34,6 +34,17 @@ SIMPLESolve::SIMPLESolve(Executioner & ex)
     _momentum_systems.push_back(&_problem.getLinearSystem(_momentum_system_numbers[system_i]));
   }
+  // and for the passive scalar equations
+  if (_has_passive_scalar_systems)
+  {
+    for (auto system_i : index_range(_passive_scalar_system_names))
+    {
+      _passive_scalar_system_numbers.push_back(
+          _problem.linearSysNum(_passive_scalar_system_names[system_i]));
+      _passive_scalar_systems.push_back(
+          &_problem.getLinearSystem(_passive_scalar_system_numbers[system_i]));
+    }
+  }
```
```suggestion
  if (_has_passive_scalar_systems)
    for (auto system_i : index_range(_passive_scalar_system_names))
    {
      _passive_scalar_system_numbers.push_back(
          _problem.linearSysNum(_passive_scalar_system_names[system_i]));
      _passive_scalar_systems.push_back(
          &_problem.getLinearSystem(_passive_scalar_system_numbers[system_i]));
    }
```

grmnptr commented on modules/navier_stokes/src/executioners/SIMPLESolve.C:
```diff
@@ -365,6 +376,48 @@ SIMPLESolve::solve()
   }
 
+  // If we have passive scalar equations, solve them here. We assume the material properties in the
+  // Navier-Stokes equations do not depend on passive scalar, as they are passive, therefore we
```
```suggestion
  // Navier-Stokes equations do not depend on passive scalsar, as they are passive, therefore we
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+  // We support internal boundaries too so we have to make sure the normal points always outward
+  const auto factor = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM) ? 1.0 : -1.0;
+
+  return boundary_value_matrix_contrib * factor * _face_mass_flux * _current_face_area;
```
Maybe dividing with the face density here should be good.

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVScalarTransportPhysics.C:
```diff
@@ -0,0 +1,232 @@
+  return params;
+}
+
+WCNSLinearFVScalarTransportPhysics::WCNSLinearFVScalarTransportPhysics(
```
Very nice!

grmnptr reviewed (changes requested):
Nice work, just a few comments.

GiudGiud commented on modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:
```diff
@@ -0,0 +1,23 @@
+
+This kernel adds the contributions of the scalar advection term to the matrix and right hand side of the scalar equation system for the finite volume SIMPLE segregated solver [SIMPLE.md].
+
+This term is described by $\nabla \cdot \left(\rho\vec{u} c_p T \right)$ present in the scalar equation conservation for an incompressible/weakly-compressible formulation.
```
well that is an important detail I missed. I ll have to rework this

GiudGiud committed:
Address Peter's review:
- use face velocities to advect scalars
- use solverVariables for physics action on add_variable task

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2575213/), step Clang format on 98f721d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cedfd055b01b8ccebc50e5d10a89847caa8e24ae`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2575264/), step Clang format on 1d8c3fb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cedfd055b01b8ccebc50e5d10a89847caa8e24ae`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2575417/), step Clang format on d31db5c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29156/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cedfd055b01b8ccebc50e5d10a89847caa8e24ae`

GiudGiud commented:
should be fixed

grmnptr commented on modules/navier_stokes/doc/content/source/linearfvkernels/LinearFVScalarAdvection.md:
```diff
@@ -0,0 +1,23 @@
+
+where $C_{if}$ is the face value of the scalar concentration. An interpolation scheme (e.g. upwind) can be used to compute the face value. This kernel adds the face contribution for each face $f$ to the right hand side and matrix.
+
+The face mass flux $(\vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```
```suggestion
The volumetric face flux $(\vec{u}\cdot \vec{n})_{RC}$ is provided by the [RhieChowMassFlux.md] object which uses pressure
```

grmnptr commented on modules/navier_stokes/doc/content/source/physics/WCNSLinearFVScalarTransportPhysics.md:
```diff
@@ -0,0 +1,43 @@
+- $\phi_i$ is the i-th scalar quantity
+- \mathbf{v} is the advecting velocity
+- $k_i$ the i-th scalar diffusivity
+- $Q_i$ is the i-th precursor source
```
```suggestion
- $Q_i$ is the i-th scalar source
```

grmnptr commented on modules/navier_stokes/include/userobjects/RhieChowMassFlux.h:
```diff
@@ -44,8 +44,10 @@ class RhieChowMassFlux : public GeneralUserObject,
-  /// Get the face velocity (used in advection terms)
+  /// Get the face velocity times density (used in advection terms)
   Real getMassFlux(const FaceInfo & fi) const;
+  /// Get the face velocity (used in advection terms)
```
```suggestion
  /// Get the volumetric face flux (used in advection terms)
```

grmnptr commented on modules/navier_stokes/include/userobjects/RhieChowMassFlux.h:
```diff
@@ -44,8 +44,10 @@ class RhieChowMassFlux : public GeneralUserObject,
+  /// Get the face velocity times density (used in advection terms)
   Real getMassFlux(const FaceInfo & fi) const;
+  /// Get the face velocity (used in advection terms)
+  Real getFaceVelocity(const FaceInfo & fi) const;
```
```suggestion
  Real getVolumetricFaceFlux(const FaceInfo & fi) const;
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+  : LinearFVFluxKernel(params),
+    _mass_flux_provider(getUserObject<RhieChowMassFlux>("rhie_chow_user_object")),
+    _advected_interp_coeffs(std::make_pair<Real, Real>(0, 0)),
+    _face_velocity(0.0)
```
```suggestion
    _volumetric_face_flux(0.0)
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+Real
+LinearFVScalarAdvection::computeElemMatrixContribution()
+{
+  return _advected_interp_coeffs.first * _face_velocity * _current_face_area;
```
```suggestion
  return _advected_interp_coeffs.first * _volumetric_face_flux * _current_face_area;
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+Real
+LinearFVScalarAdvection::computeNeighborMatrixContribution()
+{
+  return _advected_interp_coeffs.second * _face_velocity * _current_face_area;
```
```suggestion
  return _advected_interp_coeffs.second * _volumetric_face_flux * _current_face_area;
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+  // We support internal boundaries too so we have to make sure the normal points always outward
+  const auto factor = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM) ? 1.0 : -1.0;
+
+  return boundary_value_matrix_contrib * factor * _face_velocity * _current_face_area;
```
```suggestion
  return boundary_value_matrix_contrib * factor * _volumetric_face_flux * _current_face_area;
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+  const auto factor = (_current_face_type == FaceInfo::VarFaceNeighbors::ELEM ? 1.0 : -1.0);
+
+  const auto boundary_value_rhs_contrib = adv_bc->computeBoundaryValueRHSContribution();
+  return -boundary_value_rhs_contrib * factor * _face_velocity * _current_face_area;
```
```suggestion
  return -boundary_value_rhs_contrib * factor * _volumetric_face_flux * _current_face_area;
```

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVScalarAdvection.C:
```diff
@@ -0,0 +1,103 @@
+
+  // Caching the velocity on the face which will be reused in the advection term's matrix and right
+  // hand side contributions
+  _face_velocity = _mass_flux_provider.getFaceVelocity(*face_info);
```
```suggestion
  _volumetric_face_flux = _mass_flux_provider.getVolumetricFaceFlux(*face_info);
```

grmnptr commented on modules/navier_stokes/src/userobjects/RhieChowMassFlux.C:
```diff
@@ -234,6 +234,18 @@ RhieChowMassFlux::getMassFlux(const FaceInfo & fi) const
 }
 
+Real
+RhieChowMassFlux::getFaceVelocity(const FaceInfo & fi) const
```
```suggestion
RhieChowMassFlux::getVolumetricFaceFlux(const FaceInfo & fi) const
```

grmnptr commented on modules/navier_stokes/src/userobjects/RhieChowMassFlux.C:
```diff
@@ -234,6 +234,18 @@ RhieChowMassFlux::getMassFlux(const FaceInfo & fi) const
+                                /*correct_skewness=*/false,
+                                &fi.elem()};
+  const Real face_rho = _rho(face_arg, Moose::currentState());
+  return _face_mass_flux.evaluate(&fi) / face_rho;
```
```suggestion
  return _face_mass_flux[fi.id()] / face_rho;
```

grmnptr reviewed (changes requested):
Nice, almost there!

GiudGiud committed:
Address Peter's review

- face velocity -> volumetric face flux due to A term

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

Apply suggestions from code review

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

grmnptr reviewed (approved):
Nice!

GiudGiud commented:
Thanks for the review!
@tanoret fyi you may want to rebase to get the latest version of the solve object

GiudGiud commented:
Forgot to press enter:
test failures are unrelated. A distributed p2 failure in OpenMPI in unrelated tests, bluecrab not building due to an update in lowerD elements for Griffin that is unrelated as well
